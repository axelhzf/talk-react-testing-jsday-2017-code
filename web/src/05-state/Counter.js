import React from 'react';

export class Counter extends React.Component {
  state = {
    value: 0
  };

  componentDidMount() {
    this.interval = setInterval(() => {
      this.setState({ value: this.state.value + 1 });
    }, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  render() {
    return <div>{this.state.value}</div>;
  }
}
