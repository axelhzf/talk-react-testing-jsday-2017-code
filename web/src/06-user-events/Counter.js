import React from 'react';

export class Counter extends React.Component {

  state = {
    value: 0
  }

  componentDidMount () {
    this.startInterval()
  }

  componentWillUnmount () {
    this.stopInterval();
  }

  resetInterval() {
    this.setState({ value: 0 });
    this.stopInterval();
    this.startInterval();
  }

  startInterval() {
    this.interval = setInterval(() => {
      this.setState({ value: this.state.value + 1 });
    }, 1000);
  }

  stopInterval() {
    if (this.interval) clearInterval(this.interval);
  }

  handleResetClick = () => this.resetInterval();

  handleKeyPress = (e) => {
    if (e.key === 'r') {
      this.resetInterval();
    }
  }

  render () {
    return (
      <div className="container" tabIndex={0} onKeyPress={this.handleKeyPress}>
        <div>{this.state.value}</div>
        <button onClick={this.handleResetClick}>Reset</button>
      </div>
    )
  }
}