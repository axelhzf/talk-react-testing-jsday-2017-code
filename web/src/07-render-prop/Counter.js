import React from 'react';
import PropTypes from 'prop-types';

export class Counter extends React.Component {
  static propTypes = {
    render: PropTypes.func
  };

  state = {
    value: 0
  };

  componentDidMount() {
    this.startInterval();
  }

  componentWillUnmount() {
    this.stopInterval();
  }

  resetInterval() {
    this.setState({ value: 0 });
    this.stopInterval();
    this.startInterval();
  }

  startInterval() {
    this.interval = setInterval(() => {
      this.setState({ value: this.state.value + 1 });
    }, 1000);
  }

  stopInterval() {
    if (this.interval) clearInterval(this.interval);
  }

  handleResetClick = () => this.resetInterval();

  render() {
    const { render } = this.props;
    return (
      <div>
        {render({
          value: this.state.value,
          reset: this.handleResetClick
        })}
      </div>
    );
  }
}
