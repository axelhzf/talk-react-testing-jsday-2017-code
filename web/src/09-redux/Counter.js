import React from 'react';
import { combineReducers, createStore } from 'redux';
import { connect } from 'react-redux';

function counterReducer(state = 0, action) {
  switch (action.type) {
    case 'INCREMENT':
      return state + 1;
    case 'DECREMENT':
      return state - 1;
    default:
      return state;
  }
}

export const reducer = combineReducers({
  counter: counterReducer
});

const increment = () => ({ type: 'INCREMENT' });
const decrement = () => ({ type: 'DECREMENT' });

export const createDefaultStore = () => {
  const store = createStore(
    reducer,
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  );
  return store;
};

function Counter({ value, increment, decrement }) {
  return (
    <div>
      <div>{value}</div>
      <button className="increment" onClick={increment}>
        +1
      </button>
      <button className="decrement" onClick={decrement}>
        -1
      </button>
    </div>
  );
}

const mapStateToProps = state => ({
  value: state.counter
});

export const CounterContainer = connect(mapStateToProps, {
  increment,
  decrement
})(Counter);
