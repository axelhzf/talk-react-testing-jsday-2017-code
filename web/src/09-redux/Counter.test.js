import React from 'react';
import { CounterContainer, createDefaultStore, reducer } from './Counter';
import { mount } from 'enzyme';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

describe('09-redux', () => {
  it('should render initial state', () => {
    const store = createDefaultStore();
    const wrapper = mount(
      <Provider store={store}>
        <CounterContainer />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render increment', () => {
    const store = createDefaultStore();
    const wrapper = mount(
      <Provider store={store}>
        <CounterContainer />
      </Provider>
    );
    wrapper.find('.increment').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('should render decrement', () => {
    const store = createDefaultStore();
    const wrapper = mount(
      <Provider store={store}>
        <CounterContainer />
      </Provider>
    );
    wrapper.find('.decrement').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('should update redux state using a log middleware', () => {
    const states = [];
    const logMiddleware = store => next => action => {
      const state = store.getState();
      next(action);
      const nextState = store.getState();
      states.push({ state, action, nextState });
      return nextState;
    };

    const store = createStore(reducer, applyMiddleware(logMiddleware));

    const wrapper = mount(
      <Provider store={store}>
        <CounterContainer />
      </Provider>
    );
    wrapper.find('.decrement').simulate('click');
    wrapper.find('.decrement').simulate('click');
    wrapper.find('.increment').simulate('click');

    expect(states).toEqual([
      {
        state: { counter: 0 },
        action: { type: 'DECREMENT' },
        nextState: { counter: -1 }
      },
      {
        state: { counter: -1 },
        action: { type: 'DECREMENT' },
        nextState: { counter: -2 }
      },
      {
        state: { counter: -2 },
        action: { type: 'INCREMENT' },
        nextState: { counter: -1 }
      }
    ]);
  });
});
