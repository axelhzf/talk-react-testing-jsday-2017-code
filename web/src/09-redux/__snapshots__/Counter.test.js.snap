// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`09-redux should render decrement 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(Counter)>
    <Counter
      decrement={[Function]}
      increment={[Function]}
      value={-1}
    >
      <div>
        <div>
          -1
        </div>
        <button
          className="increment"
          onClick={[Function]}
        >
          +1
        </button>
        <button
          className="decrement"
          onClick={[Function]}
        >
          -1
        </button>
      </div>
    </Counter>
  </Connect(Counter)>
</Provider>
`;

exports[`09-redux should render increment 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(Counter)>
    <Counter
      decrement={[Function]}
      increment={[Function]}
      value={1}
    >
      <div>
        <div>
          1
        </div>
        <button
          className="increment"
          onClick={[Function]}
        >
          +1
        </button>
        <button
          className="decrement"
          onClick={[Function]}
        >
          -1
        </button>
      </div>
    </Counter>
  </Connect(Counter)>
</Provider>
`;

exports[`09-redux should render initial state 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(Counter)>
    <Counter
      decrement={[Function]}
      increment={[Function]}
      value={0}
    >
      <div>
        <div>
          0
        </div>
        <button
          className="increment"
          onClick={[Function]}
        >
          +1
        </button>
        <button
          className="decrement"
          onClick={[Function]}
        >
          -1
        </button>
      </div>
    </Counter>
  </Connect(Counter)>
</Provider>
`;

exports[`09-redux should update redux state 1`] = `
"[
  {
    \\"state\\": {
      \\"counter\\": 0
    },
    \\"action\\": {
      \\"type\\": \\"DECREMENT\\"
    },
    \\"nextState\\": {
      \\"counter\\": -1
    }
  },
  {
    \\"state\\": {
      \\"counter\\": -1
    },
    \\"action\\": {
      \\"type\\": \\"DECREMENT\\"
    },
    \\"nextState\\": {
      \\"counter\\": -2
    }
  },
  {
    \\"state\\": {
      \\"counter\\": -2
    },
    \\"action\\": {
      \\"type\\": \\"INCREMENT\\"
    },
    \\"nextState\\": {
      \\"counter\\": -1
    }
  }
]"
`;
